tests/virt_autotest/sriov_network_card_pci_passthrough.pm

```diff
@@ -411,19 +420,27 @@ sub save_network_device_status_logs {
 sub post_fail_hook {
     my $self = shift;
 
-    my $log_dir = "/tmp/sriov_pcipassthru";
-
     diag("Module sriov_network_card_pci_passthrough post fail hook starts.");
-    my $vm_hostnames       = script_output "virsh list --all --name | grep sles";
-    my @vm_hostnames_array = split(/\n+/, $vm_hostnames);
-    foreach (@vm_hostnames_array)
-    {
-        save_network_device_status_logs($log_dir, $_, "post_fail_hook");
+    my $log_dir = "/tmp/sriov_pcipassthru";
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        save_network_device_status_logs($log_dir, $guest, "post_fail_hook");
     }
-
     upload_virt_logs($log_dir, "network_device_status");
     $self->SUPER::post_fail_hook;
 
+    my $downloaded_xml_dir = "/tmp/download_vm_xml";
+    #restore guests for subsequent test modules
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        remove_vm($guest);
+        if (script_run("ls $downloaded_xml_dir/$guest.xml") == 0) {
+            restore_downloaded_guests($guest, $downloaded_xml_dir);
+            record_info "Guest $guest is restored.";
+        }
+    }
+}
+
+sub test_flags {
+    return {fatal => 0};
```

Comment:
I think you'd better make sure if guest restoration fails, then fatal should be 1 instead of 0.

Reply:
test should not abort in the case of a guest restoration fails. If there is a successful one, the test should continue.

Reply:
Why test test should not abort if restoration fails ?

Reply:
You have a point. Sriov test is different from other test error handling mechanisms. If all the tests 'continue next test on failure', in the case of one of the 2 guest fails, the other guest still has a chance to run all the remaining tests; But at present, there is only sriov test 'continue on failure', so it is meaningless to continue other tests as the test is expected to abort in the next test. So we're not going to continue? Or should all tests continue on failure?

Reply:
You have a point. Sriov test is different from other test error handling mechanisms. If all the tests 'continue next test on failure', in the case of one of the 2 guest fails, the other guest still has a chance to run all the remaining tests; But at present, there is only sriov test 'continue on failure', so it is meaningless to continue other tests as the test is expected to abort in the next test. So we're not going to continue? Or should all tests continue on failure?

Reply:
What I care most is if all guests restorations fail, the entire test suite should stop or terminate. If later this sriov test is integrated into other test suite, there is no point to continue test if all guests restorations fail. Although you can make this sriov the last test module in the test, the test design should be considered carefully. Otherwise, if someone else picks up this sriov test and chain it with other test together, then the entire test will continue to proceed if all guests restorations already fail.

Reply:
that's ok to fail all test when all guest restoration fails.

Reply:
I tried to figure this out, but it is a bit complex to set 'whether fail on failure' on fly. A global variable(or a test setting which is not preferred) is needed to set in a function, then call it in test_flags() function. I am not sure it is can be accepted, or if it is worth it.

Reply:
If I remember correctly, sriov is not the first test module after guest installation finishes. If not the first, then if guest restoration fail at this sriov module, it will fail at the firt module too but the first module does not have such continue test when failing setting, then the whole test will fail on the first module, no chance to go to sriov. If it really goes here, then guest restoration should pass for at least 1 guest, right? If yes, then it should be safe to have it just as it is now. Right?

Reply:
Tried but found it is impossible to set the 'continue on failure' on fly.  test_flags() function is called prior to test_run. So we have to determine it as a static value.